### PR TITLE
A new rule that finds used underscore prefixed elements

### PR DESCRIPTION
**Pull request by webwarrior-ws**
Supersedes #591 

Fixes https://github.com/fsprojects/FSharpLint/issues/573

### CODE REVIEW

**Comment by knocte** [comment]
> This branch has conflicts that must be resolved

@webwarrior-ws bad rebase ^ ; also, please push 1 by 1

File: .github/workflows/build+test+publish.yml [hunk 1]
```diff
@@ -89,3 +87,36 @@ jobs:
...
+    - name: Run Fornax
+      run: dotnet fake build -t Docs
+    - name: Deploy (if tag)
+      if: startsWith(github.ref, 'refs/tags/')
+      uses: peaceiris/actions-gh-pages@v3
+      with:
+        personal_token: ${{ secrets.GITHUB_TOKEN  }}
+        publish_dir: ./docs/_public
+        publish_branch: gh-pages
+        force_orphan: true
```
Comment (knocte) [inline .github/workflows/build+test+publish.yml:122]: @webwarrior-ws I dont know what have you done but this diff contains so many things that don't belong to this PR; bad rebase or bad conflict resolution?

**Comment by knocte** [comment]
@webwarrior-ws I think what you did wrong here when rebasing is overlooking/disregarding the last sentence (in parentheses) here: https://github.com/nblockchain/conventions/blob/master/docs/WorkflowGuidelines.md?plain=1#L245

**Comment by webwarrior-ws** [comment]
When I rebased there were no conflicts. Then I was squashing commits and something must have gone wrong there.

**Comment by knocte** [comment]
> and something must have gone wrong there.

Exactly what I meant when I pointed to our documentation, please read this line: https://github.com/nblockchain/conventions/blob/master/docs/WorkflowGuidelines.md?plain=1#L245

File: docs/content/how-tos/rules/FL0082.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+---
+title: FL0082
+category: how-to
+hide_menu: true
+---
+
+# UsedUnderscorePrefixedElements (FL0082)
+
+*Introduced in `0.21.8`*
```
Comment (knocte) [inline docs/content/how-tos/rules/FL0082.md:9]: @webwarrior-ws please fix this to say 0.23.0

File: .github/workflows/build+test+publish.yml (hunk 1 above)
Comment (webwarrior-ws) [reply .github/workflows/build+test+publish.yml]: Rebased again. Now there are no conflicts.

File: tests/FSharpLint.Core.Tests/Rules/Conventions/UsedUnderscorePrefixedElements.fs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        ()"""
+
+        Assert.IsFalse this.ErrorsExist
+
+    [<Test>]
+    member this.``Use of underscore variable in function should not cause error``() =
+        this.Parse """
+module MyModule =
+    let MyFunc () =
+        let f _ = System.Random()"""
```
Comment (knocte) [inline tests/FSharpLint.Core.Tests/Rules/Conventions/UsedUnderscorePrefixedElements.fs:82]: @webwarrior-ws I don't think this snippet compiles? you cannot end a function with a let

File: docs/content/how-tos/rules/FL0082.md (hunk 2 above)
Comment (webwarrior-ws) [reply docs/content/how-tos/rules/FL0082.md]: updated version number to 0.23.0

File: tests/FSharpLint.Core.Tests/Rules/Conventions/UsedUnderscorePrefixedElements.fs (hunk 3 above)
Comment (webwarrior-ws) [reply tests/FSharpLint.Core.Tests/Rules/Conventions/UsedUnderscorePrefixedElements.fs]: Added `()` at the end

**Comment by knocte** [comment]
@webwarrior-ws shouldn't 06fb857b44d9fa988e377ae07b39641ec312101e's CI be green instead of red?

**Comment by webwarrior-ws** [comment]
> @webwarrior-ws shouldn't [06fb857](https://github.com/fsprojects/FSharpLint/commit/06fb857b44d9fa988e377ae07b39641ec312101e)'s CI be green instead of red?

In this commit I made tests pass, but self-check fails revealing another bug. Then I added test for this case in next commit and fix in the last commit.

**Comment by knocte** [comment]
I think the first 7 commits need to be squashed, we don't gain anything by having them split in many, and their CI statuses are all red anyway (without having a failing test).

**Comment by webwarrior-ws** [comment]
Squashed first 7 commits

**Comment by knocte** [comment]
@webwarrior-ws commits c0559b8603750039d156fffaa7a4504a4749df5d and 886f06474801034d2a925ca375dab1dfe5caa81f can be squashed properly too. Also don't forget to address last commit (make test pass -> CI green). Thanks

**Comment by knocte** [comment]
@webwarrior-ws please note: as the quickfix commits (the last three) require more discussion, I've removed them from this PR. So, once I merge it, you can create new PR for those commits (e.g. cherry-picking them from this backup branch where I pushed them: https://github.com/knocte/FSharpLint/commits/addQuickFixToRule82/ ).
